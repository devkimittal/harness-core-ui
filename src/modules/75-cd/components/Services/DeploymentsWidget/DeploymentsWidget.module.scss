/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.card {
  flex-grow: 1;
  overflow: hidden;
  border: 0px;
  box-shadow: none !important;
  padding: 0px !important;
  .text {
    font-size: 40px !important;
  }
  .tickerCard {
    min-width: 100px !important;
    background: var(--grey-50) !important;
  }
  .tickerContainerStyles {
    width: fit-content;
    .tickerValueStyles {
      font-size: var(--font-size-xsmall);
    }
  }
}

.tooltipCard {
  width: 100%;
  padding-top: var(--spacing-small) !important;
  .tooltipTimestamp {
    border-bottom: 1px solid var(--grey-200);
  }
}

.toggleBtns {
  align-items: center;
  align-self: flex-start;
  position: absolute;
  right: 0;

  .chartIcon {
    text-align: center;
    cursor: pointer;
    background: var(--grey-100) !important;
    width: var(--spacing-8);
    height: var(--spacing-6) !important;
    min-height: unset;
    padding: unset !important;
  }

  .active {
    background-color: var(--primary-2) !important;
    cursor: default;
    box-shadow: var(--button-shadow) !important;
    color: var(--primary-6) !important;
  }
}

.deploymentsChartContainer {
  margin-right: var(--spacing-medium);
  .summaryCardsContainer {
    margin-left: 20px;
    display: flex;
    .summaryCard {
      min-width: 142px;
      min-height: 76px;
      background-color: #fafbfc !important;
      margin-right: var(--spacing-4) !important;
      border-radius: 4px;
      padding-left: var(--spacing-4) !important;
      padding-right: var(--spacing-4) !important;

      .cardTitle {
        font-weight: 500;
        font-size: 12px;
        line-height: 16px;
        letter-spacing: 0.2px;
        color: var(--gray-700);
        margin-top: 12px;
      }
      .frequencyContainer {
        align-items: center;
        margin-right: var(--spacing-medium);
        .frequencyCount {
          margin-top: 7px;
          font-weight: 600;
          font-size: 24px;
          line-height: 32px;
          display: flex;
          align-items: center;
          color: black;
        }
        .groupByValue {
          padding-top: var(--spacing-small);
          margin-left: var(--spacing-xsmall) !important;
          text-transform: capitalize;
        }
      }
      .trendContainer {
        margin-top: 7px;
        line-height: 14px;
        .trend {
          margin-left: 6px;
          font-weight: 500;
          font-size: 10px;
          line-height: 14px;
          display: flex;
          align-items: center;
          letter-spacing: 0.2px;
        }
      }
    }
  }
}
