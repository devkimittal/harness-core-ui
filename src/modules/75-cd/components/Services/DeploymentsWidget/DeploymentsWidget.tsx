/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useContext, useMemo } from 'react'
import { useParams } from 'react-router-dom'
import { defaultTo } from 'lodash-es'
import type { TooltipFormatterContextObject } from 'highcharts'
import { Card, Color, Container, Layout, Text, PageError, Icon } from '@wings-software/uicore'
import { useStrings } from 'framework/strings'
import { getBucketSizeForTimeRange } from '@common/components/TimeRangeSelector/TimeRangeSelector'
import { PageSpinner } from '@common/components'
import { handleZeroOrInfinityTrend, renderTrend } from '@common/components/StackedSummaryBar/utils'
import { DeploymentsTimeRangeContext, INVALID_CHANGE_RATE, numberFormatter } from '@cd/components/Services/common'
import DeploymentsEmptyState from '@cd/icons/DeploymentsEmptyState.svg'
import { renderTooltipContent } from '@pipeline/components/LandingDashboardDeploymentsWidget/LandingDashboardDeploymentsWidget'
import { GetServiceDeploymentsInfoQueryParams, useGetServiceDeploymentsInfo } from 'services/cd-ng'
import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
import { OverviewChartsWithToggle } from '@common/components/OverviewChartsWithToggle/OverviewChartsWithToggle'
import { getReadableDateTime } from '@common/utils/dateUtils'
import css from '@cd/components/Services/DeploymentsWidget/DeploymentsWidget.module.scss'

export interface ChangeValue {
  value: string
  change: number
}
interface SummaryCardData {
  title: string
  count: string
  trend: string
}
interface Points {
  deployments: {
    failure: number
    success: number
    total: number
  }
  time: number
}

export interface DeploymentWidgetProps {
  serviceIdentifier?: string
}

const deploymentsTooltip = (currPoint: TooltipFormatterContextObject): string => {
  const custom = currPoint?.series?.userOptions?.custom
  const point: Points = custom?.[currPoint.key]
  const time = getReadableDateTime(point.time)
  let failureRate: string | number = 'Infinity'
  if (point?.deployments?.failure && point.deployments?.total) {
    failureRate = ((point.deployments.failure / point.deployments.total) * 100).toFixed(1) + '%'
  }
  if (point?.deployments?.failure === 0) {
    failureRate = '0'
  }
  return renderTooltipContent({
    time,
    failureRate,
    count: point?.deployments?.total,
    successCount: point?.deployments?.success,
    failureCount: point?.deployments?.failure
  })
}

const summaryCardRenderer = (cardData: SummaryCardData, groupByValue: string): JSX.Element => {
  return (
    <Container className={css.summaryCard} key={cardData.title}>
      <Text font={{ size: 'medium' }} color={Color.GREY_700} className={css.cardTitle}>
        {cardData.title}
      </Text>
      <Layout.Horizontal>
        <Layout.Horizontal className={css.frequencyContainer}>
          <Text color={Color.BLACK} font={{ size: 'large', weight: 'bold' }} className={css.frequencyCount}>
            {cardData.count}
          </Text>
          {cardData.title === 'Deployment Frequency' && (
            <Text color={Color.GREY_700} font={{ size: 'small', weight: 'semi-bold' }} className={css.groupByValue}>
              {`/ ${groupByValue.toLocaleLowerCase()}`}
            </Text>
          )}
        </Layout.Horizontal>
        <Container className={css.trendContainer} flex>
          {cardData.trend === INVALID_CHANGE_RATE + '%' ? (
            <Icon name={'caret-down'} color={Color.RED_500}></Icon>
          ) : isNaN(parseInt(cardData.trend)) ? (
            handleZeroOrInfinityTrend(cardData.trend, cardData.trend.includes('-') ? Color.RED_500 : Color.GREEN_500)
          ) : (
            <Container flex>
              {cardData.trend.includes('-')
                ? renderTrend(cardData.trend, Color.RED_500)
                : renderTrend(cardData.trend, Color.GREEN_500)}
            </Container>
          )}
        </Container>
      </Layout.Horizontal>
    </Container>
  )
}

export const getSummaryCardRenderers = (summaryCardsData: SummaryCardData[], groupByValue: string): JSX.Element => {
  return (
    <Container className={css.summaryCardsContainer}>
      {summaryCardsData?.map(currData => summaryCardRenderer(currData, groupByValue))}
    </Container>
  )
}

export const DeploymentsWidget: React.FC<DeploymentWidgetProps> = props => {
  const { getString } = useStrings()
  const { accountId, orgIdentifier, projectIdentifier } = useParams<ProjectPathProps>()
  const { serviceIdentifier } = props
  const { timeRange } = useContext(DeploymentsTimeRangeContext)

  const queryParams: GetServiceDeploymentsInfoQueryParams = useMemo(() => {
    return {
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier,
      serviceId: serviceIdentifier,
      startTime: timeRange?.range[0]?.getTime() || 0,
      endTime: timeRange?.range[1]?.getTime() || 0,
      bucketSizeInDays: getBucketSizeForTimeRange(timeRange?.range)
    }
  }, [accountId, orgIdentifier, projectIdentifier, serviceIdentifier, timeRange])

  const {
    loading,
    data: serviceDeploymentsInfo,
    error,
    refetch
  } = useGetServiceDeploymentsInfo({
    queryParams
  })
  const data = serviceDeploymentsInfo?.data
  const dataList = data?.serviceDeploymentList

  const summaryCardsData: SummaryCardData[] = useMemo(() => {
    return [
      {
        title: getString('deploymentsText'),
        count: numberFormatter(data?.totalDeployments),
        trend: numberFormatter(data?.totalDeploymentsChangeRate) + '%'
      },
      {
        title: getString('common.failureRate'),
        count: numberFormatter(defaultTo(data?.failureRate, 0)) + '%',
        trend: numberFormatter(data?.failureRateChangeRate) + '%'
      },
      {
        title: getString('pipeline.deploymentFrequency'),
        count: numberFormatter(defaultTo(data?.frequency, 0)),
        trend: numberFormatter(data?.frequencyChangeRate) + '%'
      }
    ]
  }, [serviceDeploymentsInfo, getString])

  const chartData = useMemo(() => {
    if (data?.serviceDeploymentList?.length) {
      const successData: number[] = []
      const failureData: number[] = []
      const custom: any = []
      data?.serviceDeploymentList.forEach(val => {
        successData.push(defaultTo(val.deployments?.success, 0))
        failureData.push(defaultTo(val.deployments?.failure, 0))
        custom.push(val)
      })
      const successArr = {
        name: 'Success',
        data: successData,
        color: 'var(--success)',
        custom
      }
      const failureArr = {
        name: 'Failed',
        data: failureData,
        color: 'var(--red-400)',
        custom
      }
      return [successArr, failureArr]
    }
  }, [serviceDeploymentsInfo])

  const DeploymentWidgetContainer: React.FC = ({ children }) => {
    return (
      <Card className={css.card}>
        <Layout.Vertical height={'100%'}>{children}</Layout.Vertical>
      </Card>
    )
  }

  if (
    loading ||
    error ||
    !serviceDeploymentsInfo?.data ||
    (serviceDeploymentsInfo.data.totalDeployments === 0 && serviceDeploymentsInfo.data.totalDeploymentsChangeRate === 0)
  ) {
    const component = (() => {
      if (loading) {
        return (
          <Container data-test="deploymentsWidgetLoader">
            <PageSpinner />
          </Container>
        )
      }
      if (error) {
        return (
          <Container data-test="deploymentsWidgetError" height={'100%'}>
            <PageError onClick={() => refetch()} />
          </Container>
        )
      }
      return (
        <Layout.Vertical height="100%" flex={{ align: 'center-center' }} data-test="deploymentsWidgetEmpty">
          <img width="150" height="100" src={DeploymentsEmptyState} style={{ alignSelf: 'center' }} />
          <Text color={Color.GREY_400} margin={{ top: 'medium' }}>
            {getString('cd.serviceDashboard.noDeployments', {
              timeRange: timeRange?.label
            })}
          </Text>
        </Layout.Vertical>
      )
    })()
    return <DeploymentWidgetContainer>{component}</DeploymentWidgetContainer>
  }

  return (
    <DeploymentWidgetContainer>
      <div className={css.deploymentsChartContainer}>
        <OverviewChartsWithToggle
          data={defaultTo(chartData, [])}
          summaryCards={getSummaryCardRenderers(summaryCardsData, 'DAY')}
          customChartOptions={{
            chart: { height: 170, spacing: [25, 0, 25, 0] },
            legend: { padding: 0 },
            tooltip: {
              useHTML: true,
              formatter: function () {
                return deploymentsTooltip(this)
              },
              backgroundColor: Color.BLACK,
              outside: true,
              borderColor: 'black'
            },
            xAxis: {
              labels: {
                formatter: function (this) {
                  let time = new Date().toLocaleDateString('en-US', { day: 'numeric', month: 'short' })
                  if (dataList?.length) {
                    const val = dataList?.[this.pos]?.time
                    time = val ? new Date(val).toLocaleDateString('en-US', { day: 'numeric', month: 'short' }) : time
                  }
                  return time
                }
              }
            },
            yAxis: {
              title: {
                text: '# of Deployments',
                style: { color: 'var(--grey-400)' }
              }
            }
          }}
        />
      </div>
    </DeploymentWidgetContainer>
  )
}
