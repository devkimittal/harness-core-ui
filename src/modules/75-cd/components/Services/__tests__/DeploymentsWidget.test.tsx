/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render, getByText, fireEvent } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import { DeploymentsWidget } from '@cd/components/Services/DeploymentsWidget/DeploymentsWidget'
import { deploymentsInfo } from '@cd/mock'
import * as cdngServices from 'services/cd-ng'

jest.mock('highcharts-react-official', () => () => <></>)

jest.spyOn(cdngServices, 'useGetServiceDeploymentsInfo').mockImplementation(() => {
  return { loading: false, error: false, data: deploymentsInfo, refetch: jest.fn() } as any
})

const getLoader = (container: HTMLElement): Element => container.querySelector('[data-test="deploymentsWidgetLoader"]')!
const getError = (container: HTMLElement): Element => container.querySelector('[data-test="deploymentsWidgetError"]')!
const getEmpty = (container: HTMLElement): Element => container.querySelector('[data-test="deploymentsWidgetEmpty"]')!
const getContent = (container: HTMLElement): Element =>
  container.querySelector('[data-test="deploymentsWidgetContent"]')!

describe('DeploymentsWidget', () => {
  test('should render DeploymentsWidget', () => {
    const { container } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <DeploymentsWidget />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })

  test('should display loading state', () => {
    jest.spyOn(cdngServices, 'useGetServiceDeploymentsInfo').mockImplementation(() => {
      return { loading: true, error: false, data: [], refetch: jest.fn() } as any
    })
    const { container } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <DeploymentsWidget />
      </TestWrapper>
    )
    expect(getLoader(container)).toBeTruthy()
    expect(getError(container)).toBeFalsy()
    expect(getEmpty(container)).toBeFalsy()
    expect(getContent(container)).toBeFalsy()
  })

  test('should display error state', () => {
    const refetch = jest.fn()
    jest.spyOn(cdngServices, 'useGetServiceDeploymentsInfo').mockImplementation(() => {
      return { loading: false, error: true, data: [], refetch } as any
    })
    const { container } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <DeploymentsWidget />
      </TestWrapper>
    )
    expect(getLoader(container)).toBeFalsy()
    expect(getError(container)).toBeTruthy()
    expect(getEmpty(container)).toBeFalsy()
    expect(getContent(container)).toBeFalsy()

    expect(getByText(document.body, 'Retry')).toBeDefined()
    fireEvent.click(getByText(document.body, 'Retry') as HTMLButtonElement)
    expect(refetch).toBeCalledTimes(1)
  })

  test('should display correct data', () => {
    jest.spyOn(cdngServices, 'useGetServiceDeploymentsInfo').mockImplementation(() => {
      return { loading: false, error: false, data: deploymentsInfo, refetch: jest.fn() } as any
    })
    const { container, queryByText } = render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <DeploymentsWidget />
      </TestWrapper>
    )
    expect(getLoader(container)).toBeFalsy()
    expect(getError(container)).toBeFalsy()
    expect(getEmpty(container)).toBeFalsy()

    expect(queryByText('deploymentsText')).toBeInTheDocument()
    expect(queryByText('57')).toBeInTheDocument()
    expect(queryByText('common.failureRate')).toBeInTheDocument()
    expect(queryByText('24.2%')).toBeInTheDocument()
    expect(queryByText('pipeline.deploymentFrequency')).toBeInTheDocument()

    expect(container).toMatchSnapshot()
  })

  test('should refetch data if time range is changed', () => {
    const refetch = jest.fn()
    jest.spyOn(cdngServices, 'useGetServiceDeploymentsInfo').mockImplementation(() => {
      return { loading: false, error: false, data: [], refetch } as any
    })
    render(
      <TestWrapper
        path="account/:accountId/cd/orgs/:orgIdentifier/projects/:projectIdentifier/services"
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy' }}
      >
        <DeploymentsWidget />
      </TestWrapper>
    )

    expect(refetch).toBeCalledTimes(0)
    // Todo - Jasmeet - update this test when date picker component is integrated in deployment widget
  })
})
