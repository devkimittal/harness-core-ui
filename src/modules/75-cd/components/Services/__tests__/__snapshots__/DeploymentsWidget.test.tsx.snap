// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeploymentsWidget should display correct data 1`] = `
<div>
  <div
    class="bp3-card bp3-elevation-0 Card--card card"
  >
    <div
      class="Layout--vertical StyledProps--main"
      style="height: 100%;"
    >
      <div
        class="deploymentsChartContainer"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-large StyledProps--main"
        >
          <div
            class="StyledProps--font StyledProps--main StyledProps--flex"
          >
            <div
              class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
            >
              <div
                class="StyledProps--font StyledProps--main summaryCardsContainer"
              >
                <div
                  class="StyledProps--font StyledProps--main summaryCard"
                >
                  <p
                    class="StyledProps--font StyledProps--main cardTitle StyledProps--font-size-medium StyledProps--color StyledProps--color-grey700"
                  >
                    deploymentsText
                  </p>
                  <div
                    class="Layout--horizontal StyledProps--main"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main frequencyContainer"
                    >
                      <p
                        class="StyledProps--font StyledProps--main frequencyCount StyledProps--color StyledProps--color-black StyledProps--font-size-large StyledProps--font-weight-bold"
                      >
                        57
                      </p>
                    </div>
                    <div
                      class="StyledProps--font StyledProps--main trendContainer StyledProps--flex"
                    >
                      <div
                        class="StyledProps--font StyledProps--main StyledProps--flex"
                      >
                        <span
                          class="bp3-icon bp3-icon-caret-up StyledProps--main StyledProps--color StyledProps--color-green500"
                          icon="caret-up"
                        >
                          <svg
                            data-icon="caret-up"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-up
                            </desc>
                            <path
                              d="M11.87 9.17s.01 0 0 0l-3.5-4C8.28 5.07 8.15 5 8 5s-.28.07-.37.17l-3.5 4a.495.495 0 00.37.83h7a.495.495 0 00.37-.83z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                        <p
                          class="StyledProps--font StyledProps--main StyledProps--font-xsmall StyledProps--color StyledProps--color-green500"
                        >
                          34.4%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="StyledProps--font StyledProps--main summaryCard"
                >
                  <p
                    class="StyledProps--font StyledProps--main cardTitle StyledProps--font-size-medium StyledProps--color StyledProps--color-grey700"
                  >
                    common.failureRate
                  </p>
                  <div
                    class="Layout--horizontal StyledProps--main"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main frequencyContainer"
                    >
                      <p
                        class="StyledProps--font StyledProps--main frequencyCount StyledProps--color StyledProps--color-black StyledProps--font-size-large StyledProps--font-weight-bold"
                      >
                        24.2%
                      </p>
                    </div>
                    <div
                      class="StyledProps--font StyledProps--main trendContainer StyledProps--flex"
                    >
                      <div
                        class="StyledProps--font StyledProps--main StyledProps--flex"
                      >
                        <span
                          class="bp3-icon bp3-icon-caret-up StyledProps--main StyledProps--color StyledProps--color-green500"
                          icon="caret-up"
                        >
                          <svg
                            data-icon="caret-up"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-up
                            </desc>
                            <path
                              d="M11.87 9.17s.01 0 0 0l-3.5-4C8.28 5.07 8.15 5 8 5s-.28.07-.37.17l-3.5 4a.495.495 0 00.37.83h7a.495.495 0 00.37-.83z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                        <p
                          class="StyledProps--font StyledProps--main StyledProps--font-xsmall StyledProps--color StyledProps--color-green500"
                        >
                          45.7%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="StyledProps--font StyledProps--main summaryCard"
                >
                  <p
                    class="StyledProps--font StyledProps--main cardTitle StyledProps--font-size-medium StyledProps--color StyledProps--color-grey700"
                  >
                    pipeline.deploymentFrequency
                  </p>
                  <div
                    class="Layout--horizontal StyledProps--main"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main frequencyContainer"
                    >
                      <p
                        class="StyledProps--font StyledProps--main frequencyCount StyledProps--color StyledProps--color-black StyledProps--font-size-large StyledProps--font-weight-bold"
                      >
                        324
                      </p>
                    </div>
                    <div
                      class="StyledProps--font StyledProps--main trendContainer StyledProps--flex"
                    >
                      <div
                        class="StyledProps--font StyledProps--main StyledProps--flex"
                      >
                        <span
                          class="bp3-icon bp3-icon-caret-up StyledProps--main StyledProps--color StyledProps--color-green500"
                          icon="caret-up"
                        >
                          <svg
                            data-icon="caret-up"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-up
                            </desc>
                            <path
                              d="M11.87 9.17s.01 0 0 0l-3.5-4C8.28 5.07 8.15 5 8 5s-.28.07-.37.17l-3.5 4a.495.495 0 00.37.83h7a.495.495 0 00.37-.83z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                        <p
                          class="StyledProps--font StyledProps--main StyledProps--font-xsmall StyledProps--color StyledProps--color-green500"
                        >
                          23.2%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="StyledProps--font StyledProps--main toggleBtns StyledProps--flex"
            >
              <button
                class="bp3-button bp3-active bp3-minimal Button--button StyledProps--font StyledProps--main chartIcon active Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-primary6"
                  data-icon="bar-chart"
                >
                  <svg
                    fill="none"
                    height="12"
                    viewBox="0 0 14 11"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M11.995 0v10.798H14V0h-2.005zM3.996 1.601v9.197h2.005V1.6H3.996zm4.003 1.601v7.596h1.998V3.202H7.999zM0 4.797v6h1.998v-6H0z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main chartIcon Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey700"
                  data-icon="line-chart"
                >
                  <svg
                    fill="none"
                    height="12"
                    viewBox="0 0 14 11"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.758 10.265L5.703 1.807l-2.799 4.1H0v-.746h2.51L6.032 0l2.076 8.546 2.81-3.765H14v.747h-2.706l-3.536 4.737z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DeploymentsWidget should render DeploymentsWidget 1`] = `
<div>
  <div
    class="bp3-card bp3-elevation-0 Card--card card"
  >
    <div
      class="Layout--vertical StyledProps--main"
      style="height: 100%;"
    >
      <div
        class="deploymentsChartContainer"
      >
        <div
          class="Layout--vertical Layout--layout-spacing-large StyledProps--main"
        >
          <div
            class="StyledProps--font StyledProps--main StyledProps--flex"
          >
            <div
              class="Layout--horizontal Layout--layout-spacing-medium StyledProps--main"
            >
              <div
                class="StyledProps--font StyledProps--main summaryCardsContainer"
              >
                <div
                  class="StyledProps--font StyledProps--main summaryCard"
                >
                  <p
                    class="StyledProps--font StyledProps--main cardTitle StyledProps--font-size-medium StyledProps--color StyledProps--color-grey700"
                  >
                    deploymentsText
                  </p>
                  <div
                    class="Layout--horizontal StyledProps--main"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main frequencyContainer"
                    >
                      <p
                        class="StyledProps--font StyledProps--main frequencyCount StyledProps--color StyledProps--color-black StyledProps--font-size-large StyledProps--font-weight-bold"
                      >
                        57
                      </p>
                    </div>
                    <div
                      class="StyledProps--font StyledProps--main trendContainer StyledProps--flex"
                    >
                      <div
                        class="StyledProps--font StyledProps--main StyledProps--flex"
                      >
                        <span
                          class="bp3-icon bp3-icon-caret-up StyledProps--main StyledProps--color StyledProps--color-green500"
                          icon="caret-up"
                        >
                          <svg
                            data-icon="caret-up"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-up
                            </desc>
                            <path
                              d="M11.87 9.17s.01 0 0 0l-3.5-4C8.28 5.07 8.15 5 8 5s-.28.07-.37.17l-3.5 4a.495.495 0 00.37.83h7a.495.495 0 00.37-.83z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                        <p
                          class="StyledProps--font StyledProps--main StyledProps--font-xsmall StyledProps--color StyledProps--color-green500"
                        >
                          34.4%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="StyledProps--font StyledProps--main summaryCard"
                >
                  <p
                    class="StyledProps--font StyledProps--main cardTitle StyledProps--font-size-medium StyledProps--color StyledProps--color-grey700"
                  >
                    common.failureRate
                  </p>
                  <div
                    class="Layout--horizontal StyledProps--main"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main frequencyContainer"
                    >
                      <p
                        class="StyledProps--font StyledProps--main frequencyCount StyledProps--color StyledProps--color-black StyledProps--font-size-large StyledProps--font-weight-bold"
                      >
                        24.2%
                      </p>
                    </div>
                    <div
                      class="StyledProps--font StyledProps--main trendContainer StyledProps--flex"
                    >
                      <div
                        class="StyledProps--font StyledProps--main StyledProps--flex"
                      >
                        <span
                          class="bp3-icon bp3-icon-caret-up StyledProps--main StyledProps--color StyledProps--color-green500"
                          icon="caret-up"
                        >
                          <svg
                            data-icon="caret-up"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-up
                            </desc>
                            <path
                              d="M11.87 9.17s.01 0 0 0l-3.5-4C8.28 5.07 8.15 5 8 5s-.28.07-.37.17l-3.5 4a.495.495 0 00.37.83h7a.495.495 0 00.37-.83z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                        <p
                          class="StyledProps--font StyledProps--main StyledProps--font-xsmall StyledProps--color StyledProps--color-green500"
                        >
                          45.7%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="StyledProps--font StyledProps--main summaryCard"
                >
                  <p
                    class="StyledProps--font StyledProps--main cardTitle StyledProps--font-size-medium StyledProps--color StyledProps--color-grey700"
                  >
                    pipeline.deploymentFrequency
                  </p>
                  <div
                    class="Layout--horizontal StyledProps--main"
                  >
                    <div
                      class="Layout--horizontal StyledProps--main frequencyContainer"
                    >
                      <p
                        class="StyledProps--font StyledProps--main frequencyCount StyledProps--color StyledProps--color-black StyledProps--font-size-large StyledProps--font-weight-bold"
                      >
                        324
                      </p>
                    </div>
                    <div
                      class="StyledProps--font StyledProps--main trendContainer StyledProps--flex"
                    >
                      <div
                        class="StyledProps--font StyledProps--main StyledProps--flex"
                      >
                        <span
                          class="bp3-icon bp3-icon-caret-up StyledProps--main StyledProps--color StyledProps--color-green500"
                          icon="caret-up"
                        >
                          <svg
                            data-icon="caret-up"
                            height="16"
                            viewBox="0 0 16 16"
                            width="16"
                          >
                            <desc>
                              caret-up
                            </desc>
                            <path
                              d="M11.87 9.17s.01 0 0 0l-3.5-4C8.28 5.07 8.15 5 8 5s-.28.07-.37.17l-3.5 4a.495.495 0 00.37.83h7a.495.495 0 00.37-.83z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </span>
                        <p
                          class="StyledProps--font StyledProps--main StyledProps--font-xsmall StyledProps--color StyledProps--color-green500"
                        >
                          23.2%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="StyledProps--font StyledProps--main toggleBtns StyledProps--flex"
            >
              <button
                class="bp3-button bp3-active bp3-minimal Button--button StyledProps--font StyledProps--main chartIcon active Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-primary6"
                  data-icon="bar-chart"
                >
                  <svg
                    fill="none"
                    height="12"
                    viewBox="0 0 14 11"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M11.995 0v10.798H14V0h-2.005zM3.996 1.601v9.197h2.005V1.6H3.996zm4.003 1.601v7.596h1.998V3.202H7.999zM0 4.797v6h1.998v-6H0z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main chartIcon Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey700"
                  data-icon="line-chart"
                >
                  <svg
                    fill="none"
                    height="12"
                    viewBox="0 0 14 11"
                    width="12"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7.758 10.265L5.703 1.807l-2.799 4.1H0v-.746h2.51L6.032 0l2.076 8.546 2.81-3.765H14v.747h-2.706l-3.536 4.737z"
                      fill="currentColor"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
